ui/src/app/job-details/common/debug-icons/debug-icons.component.ts

```diff
@@ -1,36 +1,99 @@
+import {JobLogContentsComponent} from "./log-contents/log-contents.component";
 
 @Component({
   selector: 'jm-debug-icons',
   templateUrl: './debug-icons.component.html',
   styleUrls: ['./debug-icons.component.css']
 })
-export class JobDebugIconsComponent {
+export class JobDebugIconsComponent implements OnInit {
   @Input() displayMessage: boolean;
   @Input() message: string;
   @Input() stdout: string;
   @Input() stderr: string;
   @Input() directory: string;
+  logFileData: Map<string, string> = new Map();
 
   constructor(private authService: AuthService,
+              private gcsService: GcsService,
+              public logContentsDialog: MatDialog,
               private sanitizer:DomSanitizer) {
   }
 
-  getResourceUrl(url: string): SafeUrl {
-    if (!url || !ResourceUtils.isResourceURL(url)) {
+  ngOnInit(): void {
+    if (this.authService.gcsReadAccess) {
+      if (this.stdout) {
+        this.getLogContents(this.stdout).then((value) => {
+          this.logFileData[this.getFileName(this.stdout)] = value;
+        });
+      }
+      if (this.stderr) {
+        this.getLogContents(this.stderr).then((value) => {
+          this.logFileData[this.getFileName(this.stderr)] = value;
+        });
+      }
+    }
+  }
+
+  getResourceUrl(url: string): string {
+    if (!url || !ResourceUtils.isResourceURL(url) || (this.authService.gcsReadAccess && !this.hasContents(this.getFileName(url)))) {
       return '';
     }
-    return this.sanitizer.bypassSecurityTrustResourceUrl(ResourceUtils.getResourceBrowserURL(url, this.authService.userEmail));
+    return ResourceUtils.getResourceBrowserURL(url, this.authService.userEmail);
   }
 
-  getTaskDirectory(directory): SafeUrl {
+  getDirectoryUrl(directory): SafeUrl {
     if (directory && ResourceUtils.isResourceURL(directory)) {
       return this.sanitizer.bypassSecurityTrustResourceUrl(ResourceUtils.getDirectoryBrowserURL(directory, this.authService.userEmail));
     }
     return '';
   }
+
+  hasContents(fileName: string): boolean {
+    return Object.keys(this.logFileData).includes(fileName) && this.logFileData[fileName] != '';
+  }
+
+  showOrLinkTo(e: MouseEvent, url: string): void {
+    e.stopPropagation();
+    if (this.hasContents(this.getFileName(url))) {
+      this.logContentsDialog.open(JobLogContentsComponent, {
+        disableClose: false,
+        data: {
+          logName: this.getFileTitle(url),
+          logContents: this.logFileData[this.getFileName(url)],
+          logLink: this.getResourceUrl(url)
+        }
+      });
+    } else if (url) {
+      window.open(this.getResourceUrl(url));
+    }
+  }
+
+  private async getLogContents(url: string): Promise<string> {
+    const bucket = ResourceUtils.getResourceBucket(url);
+    const object = ResourceUtils.getResourceObject(url);
+    return await this.gcsService.readObject(bucket, object)
+      .then(data => data);
+  }
+
+  private getFileName(filePath: string): string {
```

Comment:
TOL is there not a library function for this?

Reply:
not that I could find